Fix filtering by NUMERIC to NUMBER comparisons

jeeminso opened the pull request:
## Summary of the changes / Why this improves CrateDB
Fixes https://github.com/crate/crate/issues/16838

## Checklist

 - [ ] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [ ] Updated documentation & `sql_features` table for user facing changes
 - [ ] Touched code is covered by tests
 - [ ] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

jeeminso commented on server/src/main/java/io/crate/types/TypeCompatibility.java:
```diff
@@ -36,18 +36,13 @@ public static DataType<?> getCommonType(DataType<?> type1, DataType<?> type2) {
-                    return null;
-                }
+        // If given types share the same base, e.g. arrays, parameter types must be compatible.
+        if (type1.getTypeParameters().isEmpty() == false || type2.getTypeParameters().isEmpty() == false) {
+            try {
+                return DataTypes.merge(type1, type2);
+            } catch (IllegalArgumentException ex) {
+                return null;
             }
-            return type1;
         }
         return convertTypeByPrecedence(type1, type2);
```
I think the logic here should be,
- if any of the two data types are parametrized, we attempt to `merge`
- if non are parametrized, `convertTypeByPrecedence`

mfussenegger reviewed (commented):
Could you please try to include info on what is the problem (including why) in the PR description and commit message and also describe the solution a bit (as to _why_ it fixes it)?

Just showing the statement it's not always clear why it is wrong, or what about it gets fixed.

jeeminso commented:
Thank you @mfussenegger, I updated the description first.

Btw, this behaviour is not limited to literals,
```
cr> create table t (a numeric(4,2), b double);
insert intCREATE OK, 1 row affected (1.650 sec)
cr> insert into t values (1.11, 1.111);
INSERT OK, 1 row affected (0.169 sec)
cr> select * from t where a = b;
+------+-------+
|    a |     b |
+------+-------+
| 1.11 | 1.111 | 
+------+-------+
SELECT 1 row in set (0.124 sec)
```
Postgres returns nothing which is also my expectation.

matriv reviewed (commented):
Thx @jeeminso!
Does this also fix the issue with `a = b` (comparing 2 columns of a table)?
If so maybe we need to a test (probably integration) for this?

jeeminso commented:
> Thx @jeeminso! Does this also fix the issue with `a = b` (comparing 2 columns of a table)? If so maybe we need to a test (probably integration) for this?

Thank you, yes, added a test and updated the release entry.

matriv reviewed (approved):
thank you!
